Topologically sort table creation statemenets

kevincianfarini opened the pull request:
Closes #2559

JakeWharton commented on sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/core/lang/util/TreeUtil.kt:
```diff
@@ -161,34 +161,64 @@ fun Collection<SqlDelightQueriesFile>.forInitializationStatements(
+}
+
+private fun <T : PsiElement, E : PsiElement> ArrayList<T>.orderStatements(
+  nameSelector: T.() -> String,
```
Selectors generally take an argument not have a receiver

kevincianfarini replied:
resolved.

JakeWharton reviewed (approved):
What happens with a cycle with this algorithm?

kevincianfarini commented:
@JakeWharton ....it hangs forever! Gross. This can be seen in `:sqldelight-compiler:integration-tests`. The following tables have a cycle. 

```sql
        CREATE TABLE team (
          name TEXT PRIMARY KEY NOT NULL,
          captain INTEGER UNIQUE NOT NULL REFERENCES player(number),
          inner_type TEXT AS Shoots.Type,
          coach TEXT NOT NULL
        );

        CREATE TABLE player (
          name TEXT NOT NULL,
          number INTEGER NOT NULL,
          team TEXT REFERENCES team(name),
          shoots TEXT AS com.squareup.sqldelight.core.integration.Shoots NOT NULL,
          PRIMARY KEY (team, number)
        );
```

I guess the algorithm for ordering statements has to be different than ordering views and trigger, because those can't contain cycles. I'll write some tests for this today.

kevincianfarini commented:
I _think_ table ordering hasn't been a problem thus far because SQLDelight's most popular use case (SQLite) is permissive about relations on tables that don't yet exist. For example, in the SQLite REPL, the following is valid. 

```
sqlite> create table parent(id integer primary key, child_id integer references child(id));
sqlite> create table child(id integer primary key, parent_id integer references parent(id));
```

In contrast, PostgreSQL errors because the relation does not yet exist. 

I can detect cycles in the relations and error with a message that instructs the user to add a foriegn key constraint after a table is created. The only problem is that SQLite's permissive nature for foreign key cycles in exactly how this is achieved, and there's no other good solution as far as I can tell. 

PostgreSQL on the other hand allows you to retroactively add FK constraints. 

```sql
ALTER TABLE child ADD CONSTRAINT fk_child_1 FOREIGN KEY (parent_id) REFERENCES parent(id);
```

Do we think this is something that should be dialect specific?

kevincianfarini commented:
I've added proper topological sorting for the create table statements. It now detects cycles and will error with a helpful message. 

For the time being, I've implemented it with a flag called `allowReferenceCycles`. If we agree that this is something that should be dialect specific, then I can hook that up and finalize this PR.

kevincianfarini commented:
Barring any possible PR comments, this is complete.

JakeWharton commented:
Will squash on green

kevincianfarini commented:
Some weird memory leak detection made the build fail. Re-triggered CI to see if it's a flake...

kevincianfarini commented:
Definitely not a flake. Don't entirely understand why removing the `java.lang.IllegalArgumentException` import triggered this...

kevincianfarini commented:
Just kidding, not some memory leak thing. This looks to be the issue. 

```
* What went wrong:
  Execution failed for task ':sqldelight-idea-plugin:runPluginVerifier'.
  > java.io.IOException: Server returned HTTP response code: 401 for URL: https://packages.jetbrains.team/maven/p/intellij-plugin-verifier/intellij-plugin-verifier/org/jetbrains/intellij/plugins/verifier-cli/maven-metadata.xml

```

kevincianfarini commented:
@JakeWharton We're green! Squash away.
